vsintegration/src/FSharp.Editor/LanguageService/SingleFileWorkspaceMap.fs

```diff
@@ -74,7 +83,47 @@ type internal SingleFileWorkspaceMap(workspace: VisualStudioWorkspace,
 
         member _.OnAfterFirstDocumentLock(_, _, _, _) = VSConstants.E_NOTIMPL
 
-        member _.OnAfterSave(_) = VSConstants.E_NOTIMPL
+        member _.OnAfterSave(docCookie) = 
+            try
+                // This causes re-analysis to happen when a F# document is saved.
+                // We do this because FCS relies on the file system and existing open documents
+                // need to be re-analyzed so the changes are propogated.
+                // We only re-analyze F# documents that are dependent on the document that was just saved.
+                // We ignore F# script documents here.
+                let vsTextBuffer = rdt4.GetDocumentData(docCookie) :?> IVsTextBuffer
+                let textBuffer = editorAdaptersFactory.GetDataBuffer(vsTextBuffer)
+                let textContainer = textBuffer.AsTextContainer()
+                let mutable workspace = Unchecked.defaultof<_>
+                if Workspace.TryGetWorkspace(textContainer, &workspace) then
+                    let solution = workspace.CurrentSolution
+                    let documentId = workspace.GetDocumentIdInCurrentContext(textContainer)
+                    match box documentId with
+                    | null -> ()
+                    | _ -> 
+                        let document = solution.GetDocument(documentId)
+                        if document.Project.Language = LanguageNames.FSharp && not document.IsFSharpScript then
+                            let openDocIds = workspace.GetOpenDocumentIds()
+                            let depProjIds = document.Project.GetDependentProjectIds().Add(document.Project.Id)
+
+                            let docIdsToReanalyze =
+                                openDocIds
+                                |> Seq.filter (fun x ->
+                                    depProjIds.Contains(x.ProjectId) && x <> document.Id &&
+                                    (
+                                        let doc = solution.GetDocument(x)
+                                        match box doc with
+                                        | null -> false
+                                        | _ -> doc.Project.Language = LanguageNames.FSharp
+                                    )
+                                )
+                                |> Array.ofSeq
```

Comment:
Is it worth ordering the documents so that visible documents are reanalyzed first? Or does Roslyn do that automatically?

Reply:
I'll see if I can. A document considered visible is sort of weird. A visible document is a document that has *ever* been visible. Meaning, if you viewed a document and then you put it behind a tab, it will still be considered a visible document.

Reply:
I can't find a good way to check if a document is actually visible or not. I think I would actually have to get the window frames per document and check if its visible which might be too much.